### PR TITLE
Add the Boo item

### PR DESCRIPTION
**Pull request by Anthcny144**
This PR adds the Boo item, which steals an opponent's item and grants temporary invincibility! The item's behavior was taken from MKDS, MK8D and MKWorld
I started working on it over a year ago and never really managed to finish it as there were a lot of situations to handle

Item specs:
- Lasts for 80 frames, same duration as Mega Mushrooms and Stars
- Stolen items priority goes to using items first then first item slot
- Targets players with a lower spot than the owner's
- Does not target players who finished the race
- In double items, the Boo is sent to second slot
- If the Boo doesn't find a target, steals a random Mushroom
- Doesn't grant off-road immunity
- Instantly steals the target
- Players in Boo go through items, players and decors (except bumper, flower and flipper assets)
- A player both in Mega Mushroom and Boo loses the destroying effect of the Mega Mushroom (doesn't break decors, doesn't damage players...)
- Animation upon being sent, stealing and coming back
- Can target a player of the same team
- Can target a player in invincibility state
- Can't steal a Boo item nor a Bullet Bill in use
- Boo doesn't give item if its owner falls off in a hole
- Steals the exact number of items the target has: if the player is trailing 2 bananas from a triple set, the Boo will steal 2 bananas
- Only 1 Boo can be in play

Technical aspect:
- Handles cases where the item to steal is used on the same frame where it should be stolen
- Handles online ping
- Displayed above the Blooper and below the POW
- Added bananeX2, carapaceX2 adn carapacerougeX2 items when the Boo steals 2 of these items
- Small code rewrite for multiple items are used

This was a pain to test every possible situation but I think the item is pretty stable overall. I hope the minimap transparency is done correctly as well

boo

### CODE REVIEW

File: scripts/mk.js [hunk 1]
```diff
@@ -18226,8 +18556,8 @@ function itemDataLength(type) {
 function resetDatas() {
 	var oPlayer = oPlayers[0];
 	var playerMapping = (course != "BB")
-	 ? ["x","y","z","speed","speedinc","heightinc","rotation","rotincdir","rotinc","drift","driftinc","driftcpt","size","tourne","tombe","arme","stash","tours","demitours","champi","etoile","megachampi","billball","place"]
-	 : ["x","y","z","speed","speedinc","heightinc","rotation","rotincdir","rotinc","drift","driftinc","driftcpt","size","tourne","tombe","arme","stash","ballons","reserve","champi","etoile","megachampi"];
+	 ? ["x","y","z","speed","speedinc","heightinc","rotation","rotincdir","rotinc","drift","driftinc","driftcpt","size","tourne","tombe","arme","stash","tours","demitours","champi","etoile","megachampi","billball","boo","place"]
+	 : ["x","y","z","speed","speedinc","heightinc","rotation","rotincdir","rotinc","drift","driftinc","driftcpt","size","tourne","tombe","arme","stash","ballons","reserve","champi","etoile","megachampi","boo"];
```
Comment (tmalahie) [inline scripts/mk.js:18584]: This list is supposed to be the same as this one https://github.com/tmalahie/mkpc/blob/820af88c26ec1aec8898124a71afed4c89a1d7d5/php/api/reload.php#L17
So if you update one you have to update the other, otherwise online mode will be broken

File: scripts/mk.js [hunk 2]
```diff
@@ -18226,8 +18556,8 @@ function itemDataLength(type) {
...
 	var playerMappingExtra = (course != "BB") ? ["finaltime"]:[];
 	var cpuMapping = playerMapping.concat("aipoint");
 	var payload = {
```
Comment (tmalahie) [inline scripts/mk.js:18587]: There's something that needs to be updated for online mode => https://github.com/tmalahie/mkpc/blob/5f307136e26f926bdfd8df3268468acfa3f4ea78/scripts/mk.js#L18702
We need to reset kart state when boo counter goes from a non-zero value to zero, otherwise the kart will stay transparent forever

File: scripts/mk.js [hunk 3]
```diff
@@ -13958,8 +14257,11 @@ function canMoveTo(iX,iY,iZ, iI,iJ, iP, iZ0) {
 						continue;
 
 					let fromSelf;
-					if (collisionTest === COL_KART && collisionPlayer === oPlayers[0])
+					if (collisionTest === COL_KART && collisionPlayer === oPlayers[0]) {
+						if (collisionPlayer.boo)
+							continue;
```
Comment (tmalahie) [inline scripts/mk.js:14272]: This will allow the player to pass through all decors, including unbreaking ones like trees, trucks... -> Is it intended? On MK8 it's not the case

File: scripts/mk.js [hunk 4]
```diff
@@ -19798,12 +20131,19 @@ function move(getId, triggered) {
...
 				kartReplaced = true;
 				delete oKart.champiType;
 				delete oKart.champior;
 				delete oKart.champior0;
 				delete oKart.boomerangArme;
 				delete oKart.boomerangStash;
+
+				// boo opacity
+				oKart.sprite[0].div.style.opacity = "";
+				oKart.driftSprite[0].div.style.opacity = "";
```
Comment (tmalahie) [inline scripts/mk.js]: When playing in 2-player (split screen multiplayer) this resets the opacity on P1 screen but not on P2 screen

Comment (tmalahie) [reply scripts/mk.js]: Also maybe in battle mode we should change opacity of the balloons? (non-mandatory it's a small detail, but in MK8 they do it)

File: scripts/mk.js [hunk 5]
```diff
@@ -2793,6 +2792,43 @@ function setStarState(oKart, duration) {
...
+		}
+	}
+}
+
+function setBooState(oKart, getId, duration) {
+	if (oKart.booGiveIn > 0)
+		return false;
+
+	if (oKart.etoile)
+		setStarState(oKart, 1);
```
Comment (tmalahie) [inline scripts/mk.js]: On MK8 you keep the star when using a boo (it's the opposite, the star takes precedence over the boo, so kart doesn't appear transparent)
I wonder if we shouldn't do the same, could be a bit frustrating to lose your star

File: scripts/mk.js [hunk 6]
```diff
@@ -8579,6 +8623,249 @@ var itemBehaviors = {
...
+							const booImg = fSprite.sprites[animIdx][0];
+
+							booImg.style.opacity = Math.min(fSprite.countdown * 0.15, 1);
+
+							booImg.gameY -= 4;
+							booImg.style.top = `${iScreenScale * booImg.gameY}px`;
+						}
+
+						// laughs at target
+						else if (checkingTarget && animIdx !== null && !fSprite.cancel) {
```
Comment (tmalahie) [inline scripts/mk.js:8767]: What's `fSprite.cancel`?

File: scripts/mk.js [hunk 7]
```diff
@@ -17750,12 +18080,12 @@ function dropBoxDecorLoot(obj, pos, distrib) {
 	}
 }
 
-function touche_asset(aPosX,aPosY,aPosZ, iX,iY) {
+function touche_asset(aPosX,aPosY,aPosZ, iX,iY, isBoo) {
```
Comment (tmalahie) [inline scripts/mk.js]: Not a fan of this extra isBoo parameter, can we move the `oKart.boo > 0` check as an extra check on the output of the function?

Comment (tmalahie) [reply scripts/mk.js]: Also I see we're transparent to flippers from waluigi pinball, on MK8 it's not the case

**Review by tmalahie — COMMENTED** [review]
Nice item addition! With your profile picture you were meant to implement this feature 😆
I tried to test all the edge cases I could think of with the item, I made lots of comments I'll let you take a look

File: scripts/mk.js (hunk 7 above)
Comment (Anthcny144) [reply scripts/mk.js]: Oops that was not intended

File: scripts/mk.js (hunk 3 above)
Comment (Anthcny144) [reply scripts/mk.js:14272]: Yes this is indented, we (I and Nodac) think this makes more sense for MKPC

File: scripts/mk.js (hunk 6 above)
Comment (Anthcny144) [reply scripts/mk.js:8767]: It's a check to keep track whether the target still has a stealable item. This is to prevent the item being both used and stolen on the same frame (in local) and to handle ping (in online)
I implemented this feature but removed it for some reason, I added it back in [this commit](https://github.com/tmalahie/mkpc/pull/96/commits/1bcf2cba910c0be372f9511ba17813c84567b061). I also made it so the Boo doesn't show up on the target if the Boo is canceled

File: scripts/mk.js [hunk 8]
```diff
@@ -8709,8 +8709,11 @@ var itemBehaviors = {
 					case 0: // anim on owner (goes up) and steal target (laughs)
 						// steal target, checks if target still has an item
 						// this should handle online ping and target using their items on the frame Boo tries stealing them
-						if (checkingTarget && fSprite.countdown === 1 && !(oKart.using.length === 0 && !oKart.arme)) {
-							const targetClientSide = !isOnline || (oKart.id === identifiant || oKart.controller === identifiant);
+						const stillHasItem = !(oKart.using.length === 0 && !oKart.arme);
```
Comment (tmalahie) [inline scripts/mk.js]: This doesn't work for billball, maybe add this extra check? Like you did [here](https://github.com/Anthcny144/mkpc/blob/77886661d4201f9024c63d7efb07a4215ce4ee7f/scripts/mk.js#L17964)
```suggestion
						const stillHasItem = !(oKart.using.length === 0 && !oKart.arme) && !oKart.billball;
```

File: scripts/mk.js [hunk 9]
```diff
@@ -18481,6 +18836,9 @@ function resetDatas() {
 							oKart.turnSound = undefined;
 						if ((!aTourne && oKart.tourne) || (!aBillBall && oKart.billball))
 							stopGrinding(j);
+						
+						if (!aBoo)
```
Comment (tmalahie) [inline scripts/mk.js]: It's not exactly what I meant, the way these variables are used is "if value was non-zero before, and value is zero after, reset kart state"
So something like this basically
```suggestion
						if (aBoo && !oKart.boo)
```

Comment (tmalahie) [reply scripts/mk.js]: The way you wrote it kinda works but it's a bit inefficient, you end up resetting boo state on every sync instead of when it's necessary

File: scripts/mk.js [hunk 10]
```diff
@@ -1138,15 +1136,16 @@ function setPlanPos(frameState, lMap) {
 				var iCharR = oKart.ref.billball ? 1.5:oKart.size;
 				var iCharW = Math.round(iCharWidth*iCharR);
 				iPlanCharacters[i].style.width = iCharW +"px";
-				posImg(iPlanCharacters[i], oKart.x,oKart.y,oKart.rotation-oKart.tourne*360/21, iCharW, iMapW);
+				const booOpacity = oKart.ref.boo ? 0.5 : 1;
+				posImg(iPlanCharacters[i], oKart.x,oKart.y,oKart.rotation-oKart.tourne*360/21, iCharW, iMapW).style.opacity = booOpacity;
```
Comment (tmalahie) [inline scripts/mk.js:1144]: This will cause an issue in battle mode, when a player is eliminated from the game we lower its opacity [here](https://github.com/Anthcny144/mkpc/blob/77886661d4201f9024c63d7efb07a4215ce4ee7f/scripts/mk.js#L1125), but with this line it will reset it back to 1

File: scripts/mk.js [hunk 11]
```diff
@@ -19826,6 +20190,7 @@ function move(getId, triggered) {
 					if (oKart.etoile)
 						oKart.sprite[i].img.src = getSpriteSrc(oKart.personnage);
 				}
+				booSetOpacity(oKart, false);
```
Comment (tmalahie) [inline scripts/mk.js]: Huge edge case but this can break the ghost replay mode: if a ghost falls in a hole after finishing the race his opacity gets reset to 1

Comment (tmalahie) [reply scripts/mk.js]: https://github.com/user-attachments/assets/a2b3d5b5-4763-44bf-ac87-158a476f6853

Comment (tmalahie) [reply scripts/mk.js]: I think the simplest fix is to call `resetBooState` instead of `booSetOpacity`, just make sure to remove the previous `oKart.boo = 0` [here](https://github.com/Anthcny144/mkpc/blob/77886661d4201f9024c63d7efb07a4215ce4ee7f/scripts/mk.js#L20158)

**Review by tmalahie — COMMENTED** [review]
I checked again with your last changes, way more stable now 🙂 just a few other issues and we'll be good hopefully 🙌

File: scripts/mk.js [hunk 12]
```diff
@@ -8557,6 +8623,259 @@ var itemBehaviors = {
...
+								consumeItem(i);
+								delete oKart.champior;
+							}
+
+							if (targetClientSide)
+								syncItems.push(fSprite);
+						}
+
+						// if target doesn't have an item anymore, cancel Boo
+						else if (checkingTarget && fSprite.countdown === 1 && !stillHasItem && targetClientSide) {
```
Comment (tmalahie) [inline scripts/mk.js:8750]: stillHasItem was renamed in https://github.com/tmalahie/mkpc/pull/96/changes/8d18ccb1fabeec41315b36060d2ae5586c80a09d
```suggestion
						else if (checkingTarget && fSprite.countdown === 1 && !isStillStealable && targetClientSide) {
```

File: scripts/mk.js [hunk 13]
```diff
@@ -1136,7 +1140,7 @@ function setPlanPos(frameState, lMap) {
 				var iCharR = oKart.ref.billball ? 1.5:oKart.size;
 				var iCharW = Math.round(iCharWidth*iCharR);
 				iPlanCharacters[i].style.width = iCharW +"px";
-				const booOpacity = oKart.ref.boo ? 0.5 : 1;
+				const booOpacity = isBattleLose ? battleLoseOpacity : (oKart.ref.boo ? 0.5 : 1);
```
Comment (tmalahie) [inline scripts/mk.js:1143]: This will cause a regression in time trial mode, a ghost icon will no longer be transparent

File: scripts/mk.js (hunk 3 above)
Comment (tmalahie) [reply scripts/mk.js:14272]: I re-reviewed, this condition only applies to the player, so CPUs don't pass through decors, which is slightly illogical, what do you think?

File: scripts/mk.js [hunk 14]
```diff
@@ -19215,7 +19570,7 @@ function move(getId, triggered) {
 						break;
 					case "pointers":
 						if (!isCustom) decorType = 'assets/pivothand';
-						if (hittable) {
+						if (hittable && !isBoo) {
```
Comment (tmalahie) [inline scripts/mk.js:19573]: Maybe make the same change [here](https://github.com/Anthcny144/mkpc/blob/2607256991c8be9485c0e94d539424585e0adcad/scripts/mk.js#L19653), otherwise a player can lose its trailing items when touching oils.

**Review by tmalahie — COMMENTED** [review]
Found a few other issues, hopefully the last ones 🙌
(... I think I already said that last time 😅)

File: scripts/mk.js (hunk 14 above)
Comment (Anthcny144) [reply scripts/mk.js:19583]: I noticed the condition for dropping items out of crate decors had a bug: they would not drop an item if they were destroyed by an online CPU (since the code was only checking for `oPlayers[0]` and not the ID of the kart hoster). I also fixed the other bug mentionned where CPUs were not going through decors in Boo state
I also made a small tweak for asset decors: flippers don't damage players in Boo state anymore (this was an oversight) but the bump still applies, so basically when a Boo player hits an asset:
Oils: no damage
Bumpers: same behavior
Flippers: no damage, but still get bumped
Clock hands: no damage, go through
Flowers: same behavior, still slowed down
